chromedp_test.go

```diff
@@ -642,30 +643,68 @@ func TestDownloadIntoDir(t *testing.T) {
 	ctx, cancel := testAllocate(t, "")
 	defer cancel()
 
+	// Wrap the context into a timeouted context, to ensure test with not run endless if download fails
+	// A second should be enough for this mocked download to finish, even on slower machines
+	ctx, cancel = context.WithTimeout(ctx, 1*time.Second)
+	defer cancel()
+
 	dir := t.TempDir()
 
+	downloadContents := "some binary data"
+
 	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		switch r.URL.Path {
 		case "/data.bin":
 			w.Header().Set("Content-Type", "application/octet-stream")
-			fmt.Fprintf(w, "some binary data")
+			fmt.Fprintf(w, downloadContents)
 		default:
 			w.Header().Set("Content-Type", "text/html")
 			fmt.Fprintf(w, `go <a id="download" href="/data.bin">download</a> stuff/`)
 		}
 	}))
 	defer s.Close()
 
+	done := make(chan string, 1)
+	ListenTarget(ctx, func(v interface{}) {
+		if ev, ok := v.(*browser.EventDownloadProgress); ok {
+			if ev.State == browser.DownloadProgressStateCompleted {
+				done <- ev.GUID
+				close(done)
+			}
+		}
+	})
+
 	if err := Run(ctx,
 		Navigate(s.URL),
-		page.SetDownloadBehavior(page.SetDownloadBehaviorBehaviorAllow).WithDownloadPath(dir),
+		browser.SetDownloadBehavior(browser.SetDownloadBehaviorBehaviorAllowAndName).WithDownloadPath(dir).WithEventsEnabled(true),
 		Click("#download", ByQuery),
 	); err != nil {
 		t.Fatal(err)
 	}
 
-	// TODO: wait for the download to finish, and check that the file is in
-	// the directory.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Download timed out")
+	case guid := <-done:
+		path := filepath.Join(dir, guid)
+
+		if _, err := os.Stat(path); err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				t.Fatal("Downloaded file does not exist")
+			}
+
+			t.Fatal("Unknown error while checking downloaded file")
+		}
+
+		fileContents, err := os.ReadFile(path)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if string(fileContents) != downloadContents {
+			t.Fatalf("Expected content of file to be '%s', but got '%s'", downloadContents, string(fileContents))
+		}
```

Comment:
I think we could trust that the browser will save the file correctly. To simplify the test code, let's remove the checking of the file content.

Reply:
Fully agree, changed that!